python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py

```diff
@@ -0,0 +1,276 @@
+from datetime import datetime
+from typing import TYPE_CHECKING, AbstractSet, Mapping, NamedTuple, Optional
+
+from typing_extensions import TypeAlias
+
+from dagster import _check as check
+from dagster._core.definitions.asset_subset import AssetSubset, ValidAssetSubset
+from dagster._core.definitions.events import AssetKey, AssetKeyPartitionKey
+from dagster._utils.cached_method import cached_method
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.base_asset_graph import BaseAssetGraph
+    from dagster._core.definitions.data_version import CachingStaleStatusResolver
+    from dagster._core.definitions.definitions_class import Definitions
+    from dagster._core.definitions.partition import PartitionsDefinition
+    from dagster._core.instance import DagsterInstance
+    from dagster._utils.caching_instance_queryer import CachingInstanceQueryer
+
+
+class TemporalContext(NamedTuple):
+    """TemporalContext represents an effective time, used for business logic, and last_event_id
+    which is used to identify that state of the event log at some point in time. Put another way,
+    the value of a TemporalContext represents a point in time and a snapshot of the event log.
+
+    Effective time: This is the effective time of the computation in terms of business logic,
+    and it impacts the behavior of partioning and partition mapping. For example, if an
+    if you get the "last" partition window of a given partitions definition, it is with
+    respect to the effective time.
+
+    Last event id: Our event log has a monotonic increasing event id. This is used to
+    cursor the event log. This event_id is also propogated to derived tables to indicate.
+    This allows us to query the state of the event log at a given point in time.
+
+    Note that insertion time of the last_event_id is not the same as the effective time.
+
+    A last_event_id of None indicates that the reads will be volatile will immediately
+    reflect any subsequent writes.
+    """
+
+    effective_dt: datetime
+    last_event_id: Optional[int]
+
+
+class _AssetSliceCompatibleSubset(ValidAssetSubset): ...
+
+
+PartitionKey: TypeAlias = Optional[str]
+AssetPartition: TypeAlias = AssetKeyPartitionKey
```

Comment:
Naming opinion: Having one of these qualified with `Key` and the other not is IMO a bit confusing, since they are both actually keys. If you want to lose `AssetKeyPartitionKey`, IMO should be either:

- `Partition`
- `AssetPartition`

or

- `PartitionKey`
- `AssetPartitionKey`

Reply:
I like `asset_partition_key` and it also matches what we do with `asset_check_key`/`AssetCheckKey`

Reply:
K i'm sold. Would like consensus from @clairelin135 @OwenKephart @sryza before we do a mass renaming.

Reply:
I like `AssetKeyPartitionKey` -> `AssetPartitionKey`. 

A couple concerns I have a about `PartitionKey`:
- Main concern: if it's just a type alias, it's not enforced by the type system, which means we're likely to end up with 
 a mix of `str` and `PartitionKey` all over the codebase. And policing this in reviews feels like a potentially large time sync.
- If we do go this route, I think it should be `str`, not `Optional[str]`. There are some situations in which a partition key is required, and we'd need a way to express that.

Reply:
@sryza would that be a special magic key for unpartitioned assets?

> And policing this in reviews feels like a potentially large time sync.

I would not endorse policing this on PRs. Typing should catch it.

Reply:
> If we do go this route, I think it should be str, not Optional[str]. There are some situations in which a partition key is required, and we'd need a way to express that.

I'm thinking about methods like `TimeWindowPartitionsDefinition.time_window_for_partition_key`, which only make sense in the context of a partitioned asset. If we were to introduce `PartitionKey`, I would expect that this method would accept a `PartitionKey`. And I would want the type system to enforce that the partition key is not `None`.

Reply:
Ah: https://github.com/dagster-io/dagster/pull/20352 is important context here. My goal is for this layer and above to "think in partitions" even for "unpartitioned" assets. There would be some awkwardness for awhile as we straddle the lines between the two layers.

Reply:
I have a local branch I could revive that has a sentinel value instead of `None`, if you want to see how that route feels.

Reply:
> I would not endorse policing this on PRs. Pyright should catch it.

How would it catch it if it's just a type alias?

> My goal is for this layer and above to "think in partitions" even for "unpartitioned" assets.

I haven't been able to put the time into fully understanding why yet, but this direction makes my spidey sense tingle.  I see how this is useful in many contexts, but I'm worried that applying it too broadly will muddy the meaning of "partition" in a way that will eventually bleed out to users.

One direction to think about this from: in the UI, "partition key" has a very specific meaning, which currently matches its meaning in the codebase.  If we let these diverge, then anyone someone encounters the concept of "partition key" in the codebase, they need to ask "which meaning of partition key are we using here?"

Imagine this sequence of events:
- Internally, we start to model all assets as partitioned assets
- We discover some fundamental situations in which this doesn't make sense to do
- We back off from doing this
- The codebase is left in an inconsistent state where we treat conceive of partitions one way in some situations and another way in other situations

This isn't to say that I think we should avoid embarking on this quest entirely. But I wonder if there's a way to do it that protects us against this risk.  E.g. introducing a `PartitionKeyOrWholeAsset` class, starting to use it, and then renaming it to `PartitionKey` if we find out that it makes sense to use it everywhere.

Reply:
I feel quite strongly that this is correct direction.

Our codebase is replete with blocks like this:

```python

        def _get_status_for_asset_key(
            asset_key: AssetKey,
        ) -> Union[PartitionedAssetBackfillStatus, UnpartitionedAssetBackfillStatus]:
            if asset_graph.get(asset_key).is_partitioned:
                materialized_subset = self.materialized_subset.get_partitions_subset(
                    asset_key, asset_graph
                )
                failed_subset = self.failed_and_downstream_subset.get_partitions_subset(
                    asset_key, asset_graph
                )
                requested_subset = self.requested_subset.get_partitions_subset(
                    asset_key, asset_graph
                )

                # The failed subset includes partitions that failed and their downstream partitions.
                # The downstream partitions are not included in the requested subset, so we determine
                # the in progress subset by subtracting partitions that are failed and requested.
                requested_and_failed_subset = failed_subset & requested_subset
                in_progress_subset = requested_subset - (
                    requested_and_failed_subset | materialized_subset
                )

                return PartitionedAssetBackfillStatus(
                    asset_key,
                    len(self.target_subset.get_partitions_subset(asset_key, asset_graph)),
                    {
                        AssetBackfillStatus.MATERIALIZED: len(materialized_subset),
                        AssetBackfillStatus.FAILED: len(failed_subset - materialized_subset),
                        AssetBackfillStatus.IN_PROGRESS: len(in_progress_subset),
                    },
                )
            else:
                failed = bool(
                    asset_key in self.failed_and_downstream_subset.non_partitioned_asset_keys
                )
                materialized = bool(
                    asset_key in self.materialized_subset.non_partitioned_asset_keys
                )
                in_progress = bool(asset_key in self.requested_subset.non_partitioned_asset_keys)

                if failed:
                    return UnpartitionedAssetBackfillStatus(asset_key, AssetBackfillStatus.FAILED)
                if materialized:
                    return UnpartitionedAssetBackfillStatus(
                        asset_key, AssetBackfillStatus.MATERIALIZED
                    )
                if in_progress:
                    return UnpartitionedAssetBackfillStatus(
                        asset_key, AssetBackfillStatus.IN_PROGRESS
                    )
                return UnpartitionedAssetBackfillStatus(asset_key, None)
```

Their existence is far, far worse than confusion in my judgment. 

Pyright will be a big help in traversing the layers, since any attempt to pass `PartitionKey`  to a function typed as  `str` will error. 

We can also do a fully typed `PartitionKey`.

I'm wiling to do a temporary gross name as a mitigation, but this is direction we need to go.

Reply:
> Pyright will be a big help in traversing the layers, since any attempt to pass PartitionKey to a function typed as str will error.

I did not think `TypeAlias`es worked this way. But not an expert.

> Their existence is far, far worse than confusion in my judgment.

To be clear, I also hate that code, and I'm very very supportive of finding a way to improve this situation. I just think that redefining what partition means in Dagster is a very dramatic change, and we need a more global view of the implications before committing to it.

Reply:
> I did not think TypeAliases worked this way. But not an expert.

It just treats it as `Optional[str]`. If you try to pass it to a function that takes a `str` it will complain.

> I just think that redefining what partition means in Dagster is a very dramatic change.

I don't think this is that dramatic to be honest. The decision has _already_ been made because `AssetKeyPartitionKey` takes a `None` to represent an unpartitioned asset. This is just formalizing that _fait accompli_.

Reply:
> How do we feel about asset_partition_keys instead everywhere?

Fine w/ this -- "asset partitions" was always playing a bit fast and loose with words to save a few chars (imo the "AssetPartition" is really the persisted object, in the same way that the "Asset" describes something different than "AssetKey")

> if it's just a type alias, it's not enforced by the type system

I think we might actually want:

```python
from typing import NewType, Optional

PartitionKey = NewType("PartitionKey", Optional[str])
# not
# PartitionKey: TypeAlias = Optional[str]
```

as this will force us to explicitly pass in `PartitionKey`-annotated objects into any function that wants to process PartitionKeys.

Agree w/ Sandy though that we'll at least in the short term need some way to represent partition keys which are guaranteed to be non-None. Very concretely, we have classes like [AssetRecordsFilter](https://sourcegraph.com/github.com/dagster-io/dagster@3cdb48c3b32a262cda27cb6fc3cc2b2cb97a2332/-/blob/python_modules/dagster/dagster/_core/event_api.py?L248) which require sequences of `str`, and in turn the queries that are fired off in response to this expect non-None objects.

So we'd probably need a `StringPartitionKey` subclass of `PartitionKey` as well.

Reply:
Hmmm actually I think pyright is too permissive on this:

```python
from typing import NewType

PartitionKey = NewType('PartitionKey', str)


def _takes_str(string: str):
    ...

def foo() -> None: 
    _takes_str(PartitionKey('foo'))

```


This works. We might need a custom class?

Reply:
@schrockn 

I was thinking the following is the scenario we'd be more worried about (which pyright does catch):

```python
from typing import NewType

PartitionKey = NewType("PartitionKey", str)


def _takes_pk(pk: PartitionKey) -> None: ...


def foo() -> None:
    _takes_pk("foo")
```

imagining _takes_pk as something like: 

```python
def get_children_partitions(
        self,
        dynamic_partitions_store: DynamicPartitionsStore,
        current_time: datetime,
        asset_key: AssetKey,
        partition_key: PartitionKey,
    ) -> AbstractSet[AssetKeyPartitionKey]:
```

One issue there is that there are plenty of user-facing methods like `execute_in_process` which take a `partition_key` argument so we wouldn't be able to change those.

Reply:
I'm going to quick do a version that makes a completely new type to see how it feels.

Reply:
> One issue there is that there are plenty of user-facing methods like execute_in_process which take a partition_key argument so we wouldn't be able to change those.

Yes I think the normal, public API should be in raw strings regardless. However tradeoffs are different there than places where a user will be writing custom AMP rules that require deep understanding of the partitioning system.